### ISSUE TITLE
[Feature]: Add inbound packets for CAG's and Ship Orientation

### ISSUE THREAD

**Issue by TheDicko** [Simpit-team/KerbalSimpitRevamped#14]
## Is your feature request related to a problem? Please describe.

Whilst we have inbound packets for actions (brake, gear, lights etc) we do not have inbound packets for Custom Action Groups and for Ship Orientation (prograde, retrograde, normal, anti-normal etc).

### A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]

I use the outbound packets for actions to shift out a byte to a shift register and turn on the relevant LED's. This is ideal as they are not in the same series as my buttons (momentary toggle switches). This overcomes the problems that would have occurred if switches were used in the same series as a toggle switch (eg when reverting to a quicksave, it will unintentionally turn things on/off). Furthermore, its nice as you can have a controller and a keyboard that won't interfere with eachother. 

Now, the same cannot be said for CAG's and ship orientation, as Simpit currently doesn't send this data from the game to the Serial. It would be great to have this!

### How would it be best that the feature is implemented? What sort of data may need to be sent by the mod to controllers? Would it require a digital signal, or an analogue one? Maybe it needs a terminal command to help use it, or a GUI button?

Similar to the actionstatus_message. I would like to be able to register to the channels, that being CAGSTATUS_MESSAGE and ORIENTATION_MESSAGE. From there, we could perform a series of if statements, each with a bitwise operation to determine the particular CAG, or the particular Orientation, that are currently turned on. Consider the following code as an example, inside the messageHandler

`byte CAGS = msg[0];  

    if (CAGS & CAG0) { 

     // processing

    } 

    else{

     // processing

      }

`

and similarly for the Orientation

`byte orientation = msg[0]; 

    if (orientation & PROGRADE) { 

      // processing

    } 

    else{

      // processing

      }`

## Describe alternatives you've considered
No alternative

### What about the current features of the program do not quite, if at all, allow you to implement/carry out what you would like to do? 
As stated, works for other things, not yet implemented for others

### What is the nearest feature in the mod (If there is one), that is along the lines of what you are suggesting?

getting Action status data (brakes, gear etc)

**Comment by LRTNZ** [comment]
So, regarding the orientation - how exactly should this be handled?
Are you talking about the selected orientation for SAS to attempt to aim at? Or the orientation reticle
that you ship is closest to?

Custom Action Groups are something that I am fully planning to implement at some point, especially
how useful they appear to have become now that you can edit them midflight - would be silly not to
expose that feature to the users.

**Comment by TheDicko** [comment]
Apologies for late reply.

Yes, selected orientation. For example, you select "prograde" in game, or via external circuitry. On the game side, as per usual it will attempt to point your ship prograde. In addition to this, simpit ping a message on the orientation channel, where you can performe bit wise if statements to determine which way the ship is trying to be orientated. You can then further process this to do things as you like (for example , light up an led).

**Comment by TheDicko** [comment]
Please feel free to ask for any more clarification. Its something i would personally love to see implemented as it would enhance my controller in a great way, so happy to help as much as possible

**Comment by TheDicko** [comment]
Good to hear about CAG's. Look forward to what should be a great next release of simpit!

**Comment by PBechon** [comment]
Here is my proposal for 2 new messages.

For SAS_MODE_STATUS_MESSAGE

```
struct sasModeStatus{
  byte sasSelectedMode; //using  the enum AutopilotMode
  uint16_t sasModeAvailable; // each bit indicate if the corresponding SAS mode is available and can be set
}
```

The bitmask can be used like this 

```
if (msg.sasModeAvailable & 1<<AP_PROGRADE){
  // do stuff if the prograde mode is available
}
```

For the CAG_STATUS_MESSAGE

```
struct cagStatus{
  uint16_t cagStatus; // each bit indicate if the corresponding CAG is activated.
}
```


The bitmask can be used like this 

```
int i = 5
if (msg.sasModeAvailable & 1<<i){
  // do stuff if the 5th custom action group is available
}
```

**Comment by TheDicko** [comment]
Looks good instinctively. Is the << necessary?

**Comment by TheDicko** [comment]
`public void Update()
        {
            bool changed = false;

            for(int i = 0; i < 10; i++)
            {
                actionStatusesCurrent[i] = FlightGlobals.ActiveVessel.ActionGroups[ActionGroupIDs[i + 1]];

                if (actionStatusesCurrent[i] != actionStatusesPrevious[i])
                {
                    Debug.Log(String.Format("CAG"+(i+1) + ": is " +actionStatusesCurrent[i]));
                    changed = true;
                }
                actionStatusesPrevious[i] = actionStatusesCurrent[i];
            }

            if (CAGStatusChannel != null && changed)
            {
                CAGStatusChannel.Fire(OutboundPackets.CAGStatus, actionStatusesCurrent);
            }
        }`

**Comment by LRTNZ** [comment]
> 
> 
> Looks good instinctively. Is the << necessary?

Yeah, it is the way the bitwise logic comparison is being done. Actually a rather neat little way of doing it, that I will have to remember for future use!

**Comment by PBechon** [comment]
This was added by #22 and #24.

Currently kind of broken for AGExt with no fix that I can think of. Do you want to change the message to only support stock CAG (so the first 10) and remove references to AGExt or should we keep this issue open ?

**Comment by LRTNZ** [comment]
I think we should keep it open for the moment, just a bookmark for future reference when we get around to having another go at it.

**Comment by PBechon** [comment]
Fix the last remaining issue with ed1c3a2b049628e5c8223b09bff35b013740da47

I propose to keep this issue open until AGExt is updated to remove the extra logs.

**Comment by PBechon** [comment]
AGExt was fixed in the last version, compatible with KSP1.12. It was tested and now seem to work.
